### DISCUSSION TITLE
Pass down parameters across build configurations in the same build chain in TeamCity

### DISCUSSION THREAD

**Discussion by Yglioh** [nuke-build/nuke#618] [Ideas]
### Description

Build automation often requires to incorporate parameters that are dynamic as mentioned in the documentation. They can be passed via command-line arguments. This works rather well when you run such a build chain locally.

In TeamCity unfortunately it requires a lot more effort to be able to pass down/up values of dynamically assigned parameters across build configurations (BCs) and agents.

Currently, after generating the settings.kts file from a project setup with Nuke and importing it in TeamCity, this results in a build chain where the information of dynamic parameters gets lost when starting a dependent BC.

### Initial investigation

Parameters defined in a nuke build project are translated to environment variables in TeamCity (e.g. `env.paramName`). In order to **pass down** parameter values from the root BC to its dependencies, you need to define a configuration parameter in the root BC (e.g. `reverse.dep.*.env.paramName`). Any dependency BC that has an environment variable defined with code `env.paramName` will now get the root's parameter value injected.

Passing parameters **up** the chain requires less setup. Variables from dependency BCs can be accessed through `dep.*.env.paramName`.

### Suggestion

If you agree that passing parameters up en down in TeamCity should be supported as is it is now in Nuke, I believe there are two places to have a look at:

1. `TeamCityConfigurationParameter.Write(..)` There should be two parameters installed:
   - The current `env.paramName` However this one isn't shown in the prompt anymore.
   - A new one `reverse.dep.*.env.paramName` This is the one that will be shown in the prompt.
2. `ParameterService.GetParameter(..)` This section could get an additional block of logic only to be executed in case of TeamCity. 
   - First check if there is a value for `dep.*.env.paramName`   
   - Next check if there is a value for `reverse.dep.*.env.paramName`
   - Finally check if there is a value for `env.paramName`

**Comment by matkoch** [comment]
`ParameterService` is oblivious to CIs

**Comment by matkoch** [comment]
I'm pretty sure there's no way to make this work consistently.

**Comment by Yglioh** [comment]
I understand there are restrictions on integrating with any kind of CI. So setting up complex scenarios in TeamCity through scripting in Nuke has it limitations.
However not being able to pass parameters is a big bummer.

What do you mean by `ParameterService` bing oblivious to CIs? `CI.TeamCity.TeamCity.IsRunningTeamCity` is not going to work for example?

**Comment by Yglioh** [comment]
About passing down parameters, I just found a nice shortcut:

    package _Self.buildTypes

    import jetbrains.buildServer.configs.kotlin.v2019_2.*
    import jetbrains.buildServer.configs.kotlin.v2019_2.buildSteps.script
    
    object TestP : BuildType({

    name = "testP"

    params {
        text("reverse.dep.*env.linkparameter", "", label = "Test link parameter", allowEmpty = true)
        text("env.linkparameter", "%reverse.dep.*env.linkparameter%", label = "Test link parameter", display = ParameterDisplay.HIDDEN, allowEmpty = true)
    }
    steps {
        script {
            scriptContent = "echo %env.linkparameter%"
        }
    }})

Like this only code needs to be modified in the generation of TeamCity configuration.

For passing up parameters something similar is maybe also possible.

**Comment by matkoch** [comment]
In the nearest future i have no capacity to work on this. If it's important for you, then you can sign up for a sponsorship.

**Comment by Yglioh** [comment]
It's totally fine that this isn't on your roadmap currently. I think I'm gonna give it a try to implement it myself and give you a pull request to have a look at.
It's convenient that you have implemented the publish process in this project, that should make it easy to create local packages to play with.

**Comment by matkoch** [comment]
Actually that includes pull requests :)

**Comment by arkanmgerges** [comment]
@Yglioh 

Thanks a lot that you've mentioned that we can pass variables down the build chain in team city. I've tested it and it worked in team city 2024.03.2 (build 156319).

> (e.g. reverse.dep.*.env.paramName). Any dependency BC that has an environment variable defined with code env.paramName will now get the root's parameter value injected.
